File: 백준/Gold/1461. 도서관/도서관.py
import sys 
input = sys.stdin.readline 
sys.setrecursionlimit(10**6)
n, m = map(int, input().split())
books = list(map(int, input().split()))
pos = []
neg = []
count = []
for i in books:
    if i>0:
        pos.append(i)
    else:
        neg.append(abs(i))

pos.sort(reverse=True)
neg.sort(reverse=True)
#한 번에 가지고 움직일 수 있는 만큼
for i in range(len(pos)//m):
    count.append(pos[i*m]) #걸음 수 = 한 번에 움직인 수 중 더 큰 수
if len(pos)%m>0:#나누어 떨어지지 않을 경우
    count.append(pos[(len(pos)//m)*m])

for i in range(len(neg)//m):
    count.append(neg[i*m])
if len(neg)%m>0:
    count.append(neg[(len(neg)//m)*m])

count.sort()

res = count.pop()#종료했을 때
res += 2*sum(count)#왕복

print(res)

File: 백준/Gold/1461. 도서관/도서관.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

var (
	reader = bufio.NewReader(os.Stdin)
	writer = bufio.NewWriter(os.Stdout)
)

func main() {
	defer writer.Flush()
	var n, m, b int
	fmt.Fscanf(reader, "%d %d\n", &n, &m)
	var pos, neg, count = []int{}, []int{}, []int{}

	for i := 0; i < n; i++ {
		fmt.Fscanf(reader, "%d ", &b)
		if b > 0 {
			pos = append(pos, b)
		} else {
			neg = append(neg, b*(-1))
		}
	}
	sort.Sort(sort.Reverse(sort.IntSlice(pos)))
	sort.Sort(sort.Reverse(sort.IntSlice(neg)))
	for i := 0; i < int(len(pos)/m); i++ {
		count = append(count, pos[i*m])
	}
	if len(pos)%m > 0 {
		count = append(count, pos[int(len(pos)/m)*m])
	}
	for i := 0; i < int(len(neg)/m); i++ {
		count = append(count, neg[i*m])
	}
	if len(neg)%m > 0 {
		count = append(count, neg[int(len(neg)/m)*m])
	}
	sort.Ints(count)
	res := count[len(count)-1]
	for i := 0; i < len(count)-1; i++ {
		res += count[i] * 2
	}
	fmt.Fprintf(writer, "%d", res)
}
